### PR TITLE
Add filtering tasks/task instances by tags

### PR DESCRIPTION
**Pull request by mikaelfr**
This pull request adds a tagging system for tasks/task instances similarly to dag tags. This allows filtering by tags in Task Instance view and Graph view. 

A new table, task_tag, is added with TaskInstance's composite primary key as a foreign key. Although having only dag_id and task_id as foreign keys would be sufficient, SQLite doesn't support a foreign key relationship over just a part of a composite primary key. 

---

Screenshots of added UI functionality:
Filtering by tags
Task tags in tree view
Tags for each task instance
Tags in graph view
Highlighting tasks by tags in graph view

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by mikaelfr** [comment]
No, I haven't tested query counts for task instances. I can potentially see some performance degradation when you have tasks that run frequently with a few tags each. 

One thing I could do is to remove foreign keys from task_tag table and just manually join task instances and task tags on task_id and dag_id. This should keep task tags table lighter at least. What do you think?

**Comment by mikaelfr** [comment]
Another possibility is to just store the list of tags into a new column in task_instance table as a json string. It avoids having to do manual checks for orphaned tags which would probably be needed in the solution above with manual joins.

File: airflow/migrations/versions/7c1bef9cb7b9_merge_heads.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+Revision ID: 7c1bef9cb7b9
+Revises: 3c20cacc0044, 19d8a998c007
+Create Date: 2020-06-08 14:09:06.676386
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = '7c1bef9cb7b9'
+down_revision = ('3c20cacc0044', '19d8a998c007')
```
Comment (ashb) [inline airflow/migrations/versions/7c1bef9cb7b9_merge_heads.py]: There should not be a merge migration. Please rebase your change and update the `down_revision` on your new migratin instead.

File: airflow/models/baseoperator.py [hunk 2]
```diff
@@ -243,6 +243,8 @@ class derived from this one results in the creation of a task object,
     :param do_xcom_push: if True, an XCom is pushed containing the Operator's
         result
     :type do_xcom_push: bool
+    :param task_tags: List of tags used to identify this task
```
Comment (ashb) [inline airflow/models/baseoperator.py:267]:
```suggestion
    :param tags: List of tags used to identify this task
```

We don't need to say "task" here -- it's already on a task.

File: airflow/models/baseoperator.py [hunk 3]
```diff
@@ -359,6 +363,8 @@ def __init__(
         self.email = email
         self.email_on_retry = email_on_retry
         self.email_on_failure = email_on_failure
+        self._task_tags: Optional[List[str]] = []
+        self.task_tags = task_tags or []
```
Comment (ashb) [inline airflow/models/baseoperator.py]:
```suggestion
        self.task_tags = task_tags
```

The setter already handles None.

File: airflow/models/dag.py [hunk 4]
```diff
@@ -844,11 +844,12 @@ def get_task_instances(
             start_date = timezone.make_aware(
                 datetime.combine(start_date, datetime.min.time()))
 
-        tis = session.query(TaskInstance).filter(
+        tis = session.query(TaskInstance).options(joinedload(TaskInstance.task_tags)).filter(
```
Comment (ashb) [inline airflow/models/dag.py]: Make this an optional flag on this method -- the scheduler doesn't care about tags so shouldn't pay the cost of loading them.

File: airflow/models/taskinstance.py [hunk 5]
```diff
@@ -501,6 +524,13 @@ def refresh_from_task(self, task, pool_override=None):
         self.max_tries = task.retries
         self.executor_config = task.executor_config
         self.operator = task.__class__.__name__
+        # Mocks tags attribute is not iterable
```
Comment (ashb) [inline airflow/models/taskinstance.py]: Don't put support code for tests in the real class -- provide a better mock instead.

File: airflow/models/taskinstance.py [hunk 6]
```diff
@@ -501,6 +524,13 @@ def refresh_from_task(self, task, pool_override=None):
...
+        if isinstance(task.task_tags, coll.Iterable):
+            for tag in task.task_tags:
+                self.task_tags.append(
+                    TaskTag(name=tag, dag_id=self.dag_id,
+                            task_id=self.task_id, execution_date=self.execution_date)
+                )
```
Comment (ashb) [inline airflow/models/taskinstance.py]: Doesn't SQLA handle this automatically for the relationship?
```suggestion
                self.task_tags.append(TaskTag(name=tag))
```

**Review by ashb — CHANGES_REQUESTED** [review]
Please provide screenshots of the new UI functionality.

I'm not sure storing this against TI is actually the right thing to do -- the task tag is a property of the Task, not the TaskInstance. What about instead if we just looked at the tags property of `dag.tasks`.

i.e. something like `tasks = [task for dag.tasks if task.has_tag(val)]`

(With the planned "DAG versioning" where we store a different version of the serialized dag model when it changes any concern about this being out of date with latest definition goes away)

File: docs/howto/add-task-tags.rst [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+ .. Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+
+Add tags to tasks
```
Comment (ashb) [inline docs/howto/add-task-tags.rst]: I think this should be in the add-dag-tags.rst -- a separate page makes it harder to discover this feature.

File: airflow/example_dags/example_task_tags.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
+#
```
Comment (ashb) [inline airflow/example_dags/example_task_tags.py]: This file is never used by anything -- please don't add new example dags without very good reason as each file in this folder slows down our test suite.

File: airflow/migrations/versions/7c1bef9cb7b9_merge_heads.py (hunk 1 above)
Comment (mikaelfr) [reply airflow/migrations/versions/7c1bef9cb7b9_merge_heads.py]: Fixed in 648e9c2

File: airflow/models/baseoperator.py (hunk 2 above)
Comment (mikaelfr) [reply airflow/models/baseoperator.py:267]: It was originally just called `tags` but when some tests failed I found out that [ECSOperator](https://github.com/apache/airflow/blob/e2b219823b9435dcdb0956a64ee2fdbf2e475124/airflow/providers/amazon/aws/operators/ecs.py#L89-L90), [OpsgenieAlertOperator](https://github.com/apache/airflow/blob/e2b219823b9435dcdb0956a64ee2fdbf2e475124/airflow/providers/opsgenie/operators/opsgenie_alert.py#L50-L51) and [AzureContainerInstancesOperator](https://github.com/apache/airflow/blob/e2b219823b9435dcdb0956a64ee2fdbf2e475124/airflow/providers/microsoft/azure/operators/azure_container_instances.py#L90-L91) already have a `tags` parameter.

File: airflow/example_dags/example_task_tags.py (hunk 8 above)
Comment (mikaelfr) [reply airflow/example_dags/example_task_tags.py]: Fixed in 8c67348

File: docs/howto/add-task-tags.rst (hunk 7 above)
Comment (mikaelfr) [reply docs/howto/add-task-tags.rst]: Fixed in 8c67348

File: airflow/models/taskinstance.py (hunk 6 above)
Comment (mikaelfr) [reply airflow/models/taskinstance.py]: Apparently it does. Fixed in 9e683c4

File: airflow/models/taskinstance.py (hunk 5 above)
Comment (mikaelfr) [reply airflow/models/taskinstance.py]: Fixed in fb4f595

File: airflow/models/baseoperator.py (hunk 3 above)
Comment (mikaelfr) [reply airflow/models/baseoperator.py]: Fixed in e67fd78

File: airflow/models/dag.py (hunk 4 above)
Comment (mikaelfr) [reply airflow/models/dag.py]: I took a look at SQLAlchemy [documentation](https://docs.sqlalchemy.org/en/13/orm/loading_relationships.html) and modified the relationship to raise an exception on lazy loads by default. I then added explicit `joinload`s in a few specific places where we want to load task tags. So currently the tags are not loaded by default and are only loaded with a join in a few places.

**Comment by mikaelfr** [comment]
> Please provide screenshots of the new UI functionality.
I edited my original comment to include some.

> I'm not sure storing this against TI is actually the right thing to do -- the task tag is a property of the Task, not the TaskInstance. What about instead if we just looked at the tags property of `dag.tasks`.
> 
> i.e. something like `tasks = [task for dag.tasks if task.has_tag(val)]`
> 
> (With the planned "DAG versioning" where we store a different version of the serialized dag model when it changes any concern about this being out of date with latest definition goes away)

I'm not sure how difficult it would be to extract tasks from serialized dags in FAB TaskInstanceModelView and use those tasks to somehow filter task instances by tags. What I can gather from my limited experience with FAB, it seems that it's not very kind to doing things in ways differing from the norm.

**Pull request closed without merging**